File: backend/arrlink/core/matching.py
"""Rule matching: does a rule's matcher match an item's tags?

A rule matches an item if any of the item's tags satisfies the matcher. The
matched tag (and, for regex rules, the match object) drive template
placeholder resolution.
"""

from __future__ import annotations

import dataclasses
import functools
import json
import re


@functools.lru_cache(maxsize=512)
def _compiled(pattern: str) -> re.Pattern | None:
    """Compile (and cache) a rule's regex. Returns None for an invalid
    pattern -- a bad regex simply never matches, same as before. Cached
    because plan_links evaluates the same handful of rule patterns across
    every item in the library on every poll."""
    try:
        return re.compile(pattern)
    except re.error:
        return None


@dataclasses.dataclass
class RuleMatch:
    """Result of a rule matching an item."""

    tag: str
    regex_match: re.Match | None = None


def match_rule(
    match_type: str,
    match_value: str,
    item_tags: list[str],
    case_insensitive: bool = False,
) -> RuleMatch | None:
    """Return a :class:`RuleMatch` for the first item tag that satisfies the
    matcher, else None. See :func:`match_rule_all` for every satisfying tag."""
    matches = match_rule_all(match_type, match_value, item_tags, case_insensitive=case_insensitive)
    return matches[0] if matches else None


def _parse_list_value(match_value: str) -> set[str]:
    """A "list" match_value is comma-separated text a user typed in the rule editor."""
    try:
        decoded = json.loads(match_value)
    except (TypeError, ValueError):
        decoded = None
    if isinstance(decoded, list):
        return {str(v).strip() for v in decoded if str(v).strip()}
    return {s.strip() for s in match_value.split(",") if s.strip()}


def match_rule_all(
    match_type: str,
    match_value: str,
    item_tags: list[str],
    case_insensitive: bool = False,
) -> list[RuleMatch]:
    """Like :func:`match_rule`, but returns every item tag that satisfies the
    matcher instead of stopping at the first -- used to fan a single condition
    out into multiple destination links (one per matching tag).

    ``case_insensitive`` is for native-metadata conditions: Radarr/Sonarr's
    own fields (originalLanguage.name, genres, certifications, quality
    profile names) use a fixed, known Title Case convention, so comparing
    case-insensitively there is safe and avoids a silent, unexplained
    zero-match if a user types "english" instead of "English". Tag matching
    stays case-sensitive since tags are free-form user data where case can
    be meaningful."""
    if match_type == "exact":
        target = match_value.strip()
        if case_insensitive:
            target = target.casefold()
            return [RuleMatch(tag=t) for t in item_tags if t.casefold() == target]
        return [RuleMatch(tag=t) for t in item_tags if t == target]

    if match_type == "list":
        targets = _parse_list_value(match_value)
        if case_insensitive:
            targets = {t.casefold() for t in targets}
            return [RuleMatch(tag=t) for t in item_tags if t.casefold() in targets]
        return [RuleMatch(tag=t) for t in item_tags if t in targets]

    if match_type == "regex":
        pattern = _compiled(match_value)
        if pattern is None:
            return []
        out = []
        for t in item_tags:
            m = pattern.search(t)
            if m:
                out.append(RuleMatch(tag=t, regex_match=m))
        return out

    return []


@dataclasses.dataclass
class ConditionMatch:
    """One condition (identified by its category) that was evaluated and
    matched while folding a rule's condition chain."""

    category: str
    tag: str
    regex_match: re.Match | None = None


@dataclasses.dataclass
class ConditionsResult:
    """Result of folding an ordered AND/OR condition chain over an item's tags."""

    result: bool
    matched_conditions: list[ConditionMatch]
    all_matches: dict[str, list[ConditionMatch]] = dataclasses.field(default_factory=dict)


def match_conditions(
    conditions: list[dict],
    item_tags: list[str],
    native: dict[str, list[str]] | None = None,
) -> ConditionsResult:
    """Left-to-right, short-circuiting AND/OR fold over an ordered condition chain."""
    if not conditions:
        return ConditionsResult(result=False, matched_conditions=[], all_matches={})
    native = native or {}

    running = False
    matched: list[ConditionMatch] = []
    all_matches: dict[str, list[ConditionMatch]] = {}
    for i, cond in enumerate(conditions):
        join = cond.get("join")
        if i > 0:
            if join == "AND" and running is False:
                continue
            if join == "OR" and running is True:
                continue

        is_native = cond.get("source") == "native"
        values = native.get(cond["category"], []) if is_native else item_tags
        hits = match_rule_all(
            cond["match_type"], cond["match_value"], values, case_insensitive=is_native
        )
        hit = bool(hits)
        if i == 0:
            running = hit
        elif join == "AND":
            running = running and hit
        else:  # "OR"
            running = running or hit

        if hit:
            category = cond["category"]
            cms = [
                ConditionMatch(category=category, tag=m.tag, regex_match=m.regex_match)
                for m in hits
            ]
            matched.append(cms[0])
            all_matches[category] = cms

    return ConditionsResult(result=running, matched_conditions=matched, all_matches=all_matches)

File: backend-ts/src/core/matching.ts
/** Rule matching: does a rule's matcher match an item's tags? The matched tag
 * (and, for regex rules, the match object) drives template placeholder
 * resolution. Ported from core/matching.py. */

export interface RuleMatch {
  tag: string
  regexMatch: RegExpMatchArray | null
}

// 'vocabulary' is a stored-but-not-directly-matched type: expandVocabularyConditions
// rewrites it to 'list' before match_rule_all ever sees it. 'range' is for
// numeric native fields (rating, popularity, runtime) -- matchValue is
// {"min": number|null, "max": number|null} JSON, either bound omittable for
// an open-ended threshold ("rating >= 7" is {min:7,max:null}).
export type MatchType = 'exact' | 'list' | 'regex' | 'vocabulary' | 'range'
export type ConditionJoin = 'AND' | 'OR' | null

export interface RangeValue {
  min: number | null
  max: number | null
}

/** Parses a 'range' condition's matchValue; null if malformed (never matches,
 * same as an invalid regex). */
export function parseRangeValue(matchValue: string): RangeValue | null {
  try {
    const decoded: unknown = JSON.parse(matchValue)
    if (
      decoded !== null &&
      typeof decoded === 'object' &&
      !Array.isArray(decoded) &&
      ('min' in decoded || 'max' in decoded)
    ) {
      const d = decoded as { min?: unknown; max?: unknown }
      const min = d.min === null || d.min === undefined ? null : Number(d.min)
      const max = d.max === null || d.max === undefined ? null : Number(d.max)
      if ((min !== null && !Number.isFinite(min)) || (max !== null && !Number.isFinite(max))) {
        return null
      }
      if (min === null && max === null) return null // no bound at all -- never matches
      return { min, max }
    }
  } catch {
    // malformed JSON
  }
  return null
}

export interface Condition {
  category: string
  matchType: MatchType
  matchValue: string
  join: ConditionJoin
  source?: 'native' | 'tag' | null
}

// Python named groups `(?P<name>...)` / backreferences `(?P=name)` use JS
// syntax instead: `(?<name>...)` / `\k<name>`. Translate before compiling so
// existing user-authored rules keep working. Not covered: lookbehind version
// differences, some Unicode property escapes.
function translatePythonRegexSyntax(pattern: string): string {
  return pattern
    .replace(/\(\?P<([A-Za-z_][A-Za-z0-9_]*)>/g, '(?<$1>')
    .replace(/\(\?P=([A-Za-z_][A-Za-z0-9_]*)\)/g, '\\k<$1>')
}

// Compile (and cache) a rule's regex; null for an invalid pattern (never
// matches). Size-bounded LRU, mirrors Python's lru_cache(maxsize=512).
const REGEX_CACHE_MAX = 512
const regexCache = new Map<string, RegExp | null>()

function compiled(pattern: string): RegExp | null {
  const hit = regexCache.get(pattern)
  if (hit !== undefined) {
    regexCache.delete(pattern)
    regexCache.set(pattern, hit)
    return hit
  }
  let re: RegExp | null
  try {
    re = new RegExp(translatePythonRegexSyntax(pattern))
  } catch {
    re = null
  }
  regexCache.set(pattern, re)
  if (regexCache.size > REGEX_CACHE_MAX) {
    const oldest = regexCache.keys().next().value
    if (oldest !== undefined) regexCache.delete(oldest)
  }
  return re
}

/** Test helper: clears the compiled-regex cache. */
export function clearRegexCache(): void {
  regexCache.clear()
}

function parseListValue(matchValue: string): Set<string> {
  try {
    const decoded: unknown = JSON.parse(matchValue)
    if (Array.isArray(decoded)) {
      return new Set(decoded.map((v) => String(v).trim()).filter((v) => v.length > 0))
    }
  } catch {
    // not JSON -- fall through to comma-split
  }
  return new Set(
    matchValue
      .split(',')
      .map((s) => s.trim())
      .filter((s) => s.length > 0),
  )
}

/** Every item tag that satisfies the matcher -- fans one condition into multiple links.
 *
 * `caseInsensitive` is for native-metadata conditions: Radarr/Sonarr's own
 * fields (originalLanguage.name, genres, certifications, quality profile
 * names) use a fixed, known Title Case convention, so comparing
 * case-insensitively there is safe and avoids a silent, unexplained
 * zero-match if a user types "english" instead of "English". Tag matching
 * stays case-sensitive since tags are free-form user data where case can be
 * meaningful. */
export function matchRuleAll(
  matchType: string,
  matchValue: string,
  itemTags: string[],
  caseInsensitive = false,
): RuleMatch[] {
  if (matchType === 'exact') {
    const target = matchValue.trim()
    if (caseInsensitive) {
      const lower = target.toLowerCase()
      return itemTags
        .filter((t) => t.toLowerCase() === lower)
        .map((t) => ({ tag: t, regexMatch: null }))
    }
    return itemTags.filter((t) => t === target).map((t) => ({ tag: t, regexMatch: null }))
  }

  if (matchType === 'list') {
    const targets = parseListValue(matchValue)
    if (caseInsensitive) {
      const lowerTargets = new Set([...targets].map((t) => t.toLowerCase()))
      return itemTags
        .filter((t) => lowerTargets.has(t.toLowerCase()))
        .map((t) => ({ tag: t, regexMatch: null }))
    }
    return itemTags
      .filter((t) => targets.has(t))
      .map((t) => ({ tag: t, regexMatch: null }))
  }

  if (matchType === 'regex') {
    const pattern = compiled(matchValue)
    if (pattern === null) return []
    const out: RuleMatch[] = []
    for (const t of itemTags) {
      const m = t.match(pattern)
      if (m) out.push({ tag: t, regexMatch: m })
    }
    return out
  }

  if (matchType === 'range') {
    const range = parseRangeValue(matchValue)
    if (range === null) return []
    const inRange = (n: number): boolean =>
      (range.min === null || n >= range.min) && (range.max === null || n <= range.max)
    return itemTags
      .filter((t) => {
        const n = Number(t)
        return Number.isFinite(n) && inRange(n)
      })
      .map((t) => ({ tag: t, regexMatch: null }))
  }

  return []
}

/** The first item tag that satisfies the matcher, else null. */
export function matchRule(
  matchType: string,
  matchValue: string,
  itemTags: string[],
  caseInsensitive = false,
): RuleMatch | null {
  const matches = matchRuleAll(matchType, matchValue, itemTags, caseInsensitive)
  return matches.length > 0 ? matches[0] : null
}

/** One condition (identified by its category) that was evaluated and matched
 * while folding a rule's condition chain. */
export interface ConditionMatch {
  category: string
  tag: string
  regexMatch: RegExpMatchArray | null
}

export interface ConditionsResult {
  result: boolean
  matchedConditions: ConditionMatch[]
  allMatches: Record<string, ConditionMatch[]>
}

/** Left-to-right, short-circuiting AND/OR fold over an ordered condition chain. */
export function matchConditions(
  conditions: Condition[],
  itemTags: string[],
  native: Record<string, string[]> = {},
): ConditionsResult {
  if (conditions.length === 0) {
    return { result: false, matchedConditions: [], allMatches: {} }
  }

  let running = false
  const matched: ConditionMatch[] = []
  const allMatches: Record<string, ConditionMatch[]> = {}

  conditions.forEach((cond, i) => {
    if (i > 0) {
      if (cond.join === 'AND' && running === false) return
      if (cond.join === 'OR' && running === true) return
    }

    const isNative = cond.source === 'native'
    const values = isNative ? (native[cond.category] ?? []) : itemTags
    const hits = matchRuleAll(cond.matchType, cond.matchValue, values, isNative)
    const hit = hits.length > 0

    if (i === 0) {
      running = hit
    } else if (cond.join === 'AND') {
      running = running && hit
    } else {
      running = running || hit
    }

    if (hit) {
      const cms = hits.map((m) => ({
        category: cond.category,
        tag: m.tag,
        regexMatch: m.regexMatch,
      }))
      matched.push(cms[0])
      allMatches[cond.category] = cms
    }
  })

  return { result: running, matchedConditions: matched, allMatches }
}
